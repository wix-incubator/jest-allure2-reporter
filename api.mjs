export * from './dist/api/index.js';
