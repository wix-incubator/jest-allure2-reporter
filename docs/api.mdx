import DocCardList from '@theme/DocCardList';

# API

Learn about the features of `jest-allure2-reporter`.

<DocCardList />
