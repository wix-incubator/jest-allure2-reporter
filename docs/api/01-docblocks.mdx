---
sidebar_position: 2
---

# Docblocks

:::info

Docblocks cannot be applied to `describe` statements.
If you want to apply a docblock to all tests in the file,
you can put it on the first line of the file.

:::

Docblocks (JSDoc-style comments) are the least intrusive way to add metadata to your tests.
These annotations get parsed by `jest-allure2-reporter` and used as a source of information for your test reports.

## Plain comments

Plain comments act as [`@description`][1] annotations, when applied to a test case.

```javascript
/**
 * This test demonstrates the addition operation.
 */
test('should add two numbers', () => {
  expect(1 + 2).toBe(3);
});
```

For hooks, they act as [`@displayName`][2] annotations.

```javascript
/**
 * Navigate to the home page
 */
beforeEach(async () => {
  await page.goto('https://example.com');
});
```

## `@description` / `@desc`

Adds a Markdown description to a test case.

```javascript
/**
 * @description
 * This test demonstrates the addition operation.
 */
test('should add two numbers', () => {
  expect(1 + 2).toBe(3);
});
```

## `@descriptionHtml`

Adds an HTML description to a test case.

```javascript
/**
 * @descriptionHtml
 * This test demonstrates the <code>-</code> operator.
 */
test('should subtract two numbers', () => {
  expect(2 - 1).toBe(1);
});
```

## `@displayName`

Overrides test names specified in `test('...')` or `it('...')` in the test report.

```javascript
/**
 * @displayName 1 + 1 = 2
 */
test('First test', () => {
  expect(1 + 1).toBe(2);
});
```

When applied to a hook, it sets a custom display name for the hook, similar to a [plain comment][3]:

```javascript
/**
 * @displayName Custom "beforeEach" hook
 */
beforeEach(() => {
  // Hook implementation
});
```

## `@fullName`

Sets a full name for a test case, which can be used for more detailed identification.
By default, full names are also used for tracking test history across multiple runs or retries.

```javascript
/**
 * @fullName Arithmetic > Addition > Valid assertion
 */
test('First test', () => {
  expect(1 + 1).toBe(2);
});
```

## `@historyId`

Assigns a custom history ID to a test case, useful for tracking test history across multiple runs or retries.

```javascript
/**
 * @historyId HISTORY-2
 */
test('First test', () => {
  expect(2 + 2).toBe(3);
});
```

## `@issue`

Links an issue to a test case.

For an individual test:

```javascript
/**
 * @issue XMLRPC-15
 */
test('Proving the fix', () => {
  expect(1 + 1).toBe(2);
});
```

For all tests in the file, put the docblock at the very top:

```javascript
/**
 * @issue XMLRPC-15
 */

// Rest of your test file...
```

## `@owner`

Specifies the owner of a test or suite.

For an individual test:

```javascript
/**
 * @owner John Doe
 */
test('Test maintained by John', () => {
  // Test implementation
});
```

For all tests in the file:

```javascript
/**
 * @owner John Doe
 */

// Rest of your test file...
```

## `@package`

Specifies the package for a test or suite, useful for organizing tests.

For an individual test:

```javascript
/**
 * @package e2e.pragmas
 */
test('should log a message', () => {
  // Test implementation
});
```

For all tests in the file:

```javascript
/**
 * @package e2e.my-tests
 */

// Rest of your test file...
```

## `@severity`

Sets the severity level for a test or suite.

For an individual test:

```javascript
/**
 * @severity critical
 */
test('Important test', () => {
  expect(1 + 1).toBe(2);
});
```

For all tests in the file:

```javascript
/**
 * @severity critical
 */

// Rest of your test file...
```

## `@epic`, `@feature`, `@story`

Categorizes tests into epics and features for better organization.

Example:

```javascript
/**
 * @epic Arithmetic operations
 * @feature Addition
 */

// Rest of your test file...

/**
 * @story Sane assumptions
 */
test('1 + 1 should equal 2', () => {
  expect(1 + 1).toBe(2);
});
```

## `@tag`

Adds tags to a test or suite.

For an individual test:

```javascript
/**
 * @tag arithmetic, addition
 */
test('First test', () => {
  expect(1 + 1).toBe(2);
});
```

For all tests in the file:

```javascript
/**
 * @tag arithmetic
 */

// Rest of your test file...
```

## `@thread`

Specifies a custom thread for concurrent tests.
Do not use it unless you want to control tests on the [Timeline][4] manually.

For an individual test:

```javascript
/**
 * @thread IV
 */
test('Concurrent test', () => {
  expect(1 + 1).toBe(2);
});
```

For all tests in the file:

```javascript
/**
 * @thread IV
 */

// Rest of your test file...
```

## `@tms`

Links a test management system (TMS) case to a test.

For an individual test:

```javascript
/**
 * @tms TMS-1234
 */
test('should be linked to a TMS ticket', () => {
  expect(1 + 1).toBe(2);
});
```

For all tests in the file:

```javascript
/**
 * @tms TMS-1234
 */

// Rest of your test file...
```

## `@url`

Adds a custom URL link to a test or suite.

For an individual test:

```javascript
/**
 * @url https://en.wikipedia.org/wiki/Addition âž• Addition
 */
test('1 + 1 = 2', () => {
  expect(1 + 1).toBe(2);
});
```

For all tests in the file:

```javascript
/**
 * @url https://en.wikipedia.org/wiki/Arithmetic ðŸ”¢ Arithmetic
 */

// Rest of your test file...
```

## `@parentSuite`, `@suite`, `@subSuite`

Organizes tests into a hierarchical suite structure.

Example:

```javascript
/**
 * @parentSuite Custom Parent Suite
 */

// ...

/**
 * @suite Custom Suite
 * @subSuite Custom Sub-Suite
 */
test('Test with custom suite hierarchy', () => {
  // Test implementation
});
```

These docblock annotations provide a powerful way to enrich your tests with metadata, improving the organization and readability of your Allure reports. By using these annotations, you can create more informative and structured test reports that help in better understanding and maintaining your test suite.

[1]: #description--desc
[2]: #displayname
[3]: #plain-comments
[4]: #TODO
