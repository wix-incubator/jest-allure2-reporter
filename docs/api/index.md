---
id: "index"
title: "jest-allure2-reporter"
sidebar_label: "Readme"
sidebar_position: 0
custom_edit_url: null
---

