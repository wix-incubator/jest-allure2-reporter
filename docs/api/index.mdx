---
sidebar_position: 1
slug: /api
---

# Exports

The `jest-allure2-reporter` package provides several entry points to enhance your Jest testing experience with Allure reporting capabilities.

## Main Entry Point

### `jest-allure2-reporter`

The core reporter module for collecting test results and generating Allure reports.

```javascript title="jest.config.js"
/** @type {import('@jest/types').Config.InitialOptions} */
module.exports = {
  // ...other Jest configurations
  reporters: [
    'default',
    ['jest-allure2-reporter',
      /** @type {import('jest-allure2-reporter').ReporterOptions} */
      {
        // Reporter options go here
      }
    ]
  ],
};
```

For detailed configuration options, see the [Configuration Guide][1].

You can extend this class to customize the reporting behavior:

```javascript
import { execa } from 'execa';
import JestAllure2Reporter from 'jest-allure2-reporter';

class CustomizedJestAllure2Reporter extends JestAllure2Reporter {
  async onRunComplete(contexts, results) {
    await super.onRunComplete(contexts, results);
    // Illustrative example: Generate Allure report after test run
    await execa('allure', ['generate', '--clean']);
  }
}
```

The main entry point also exports the `ReporterOptions` type for TypeScript users, so you
can define presets and configurations with type safety:

```typescript
import type { ReporterOptions } from 'jest-allure2-reporter';

export const myPreset1: ReporterOptions = {
  resultsDir: 'artifacts/allure-results',
  testCase: {
    // ...
  },
};
```

## API Entry Point

### `jest-allure2-reporter/api`

This module exports functions and objects for enhancing your tests with Allure metadata and runtime capabilities.

```javascript
import {
  $Description,
  $Link,
  $Owner,
  allure,
  Attachment,
  Step
} from 'jest-allure2-reporter/api';
```

Key exports include:
- [Annotations][2]: Functions like `$Description`, `$Link`, `$Owner` for adding metadata to your tests.
- [Decorators][3]: `@Attachment` and `@Step` for enhancing class-based test drivers.
- [Runtime API][4]: The `allure` object for interacting with Allure during test execution.

If you prefer not to use imports, you can use the global typings provided by `jest-allure2-reporter/globals`.

### Global Typings

For TypeScript projects, you can add Allure types globally by including them in your `tsconfig.json`:

```js
{
  "compilerOptions": {
    "types": ["jest-allure2-reporter/globals"]
  }
}
```

Alternatively, you can import them directly somewhere in your test files:

```typescript
import 'jest-allure2-reporter/globals';
```

## Environment Packages

[Jest test environments][5] provide the context in which tests are run: `node`, `jsdom`, or custom environments.

Custom environments are crucial for `jest-allure2-reporter` as they enable the collection of detailed metadata and ensure that annotations, runtime API calls, and decorators can pass information to the reporter — usually this happens through interprocess communication (IPC) since Jest spawns separate worker processes for each test file, and there are no built-in mechanisms for sharing custom data between workers and the main process.

Although `jest-allure2-reporter` technically works with the default Jest environments, the produced reports will be less informative and lacking custom metadata. To get the most out of Allure, you should use one of the provided custom environments.

### `jest-allure2-reporter/environment-node`

⭐ **Recommended choice**.

This test environment is designed to work with Node.js tests and Allure.

```javascript
// jest.config.js
module.exports = {
  testEnvironment: 'jest-allure2-reporter/environment-node',
  // ...
};
```

### `jest-allure2-reporter/environment-jsdom`

A Jest environment for browser-like tests using jsdom, with Allure support.

```javascript
// jest.config.js
module.exports = {
  testEnvironment: 'jest-allure2-reporter/environment-jsdom',
  // ...
};
```

### `jest-allure2-reporter/environment-decorator`

A utility for creating custom Jest environments with Allure support.

```javascript title="custom-allure-environment.js"
import allureDecorator from 'jest-allure2-reporter/environment-decorator';
import JestEnvironmentCustom from './JestEnvironmentCustom';

export default allureDecorator(JestEnvironmentCustom);
```

This is useful for non-standard test environments where you need to bring Allure capabilities in.
For more information on custom environments, see the [Jest documentation][5].

[1]: config/index.mdx
[2]: 02-annotations.mdx
[3]: 03-decorators.mdx
[4]: 04-runtime-api.mdx
[5]: https://jestjs.io/docs/configuration#testenvironment-string
