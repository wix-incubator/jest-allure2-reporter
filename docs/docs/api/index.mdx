---
sidebar_position: 3
---

import DocCardList from '@theme/DocCardList';

# API

Learn about the features of `jest-allure2-reporter`.

<DocCardList />
