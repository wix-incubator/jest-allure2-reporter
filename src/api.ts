import type { IAllureRuntime } from './api/runtime';
import realm from './realms';

export * from './api/annotations';
export * from './api/decorators';

export const allure = realm.runtime as IAllureRuntime;

export type {
  AttachmentContent,
  AttachmentOptions,
  IAllureRuntime,
  ParameterOrString,
} from './api/runtime';
