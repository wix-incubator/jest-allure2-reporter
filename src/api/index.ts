import realm from '../realms';

import type { IAllureRuntime } from './runtime';

export * from './annotations';
export * from './decorators';

export const allure = realm.runtime as IAllureRuntime;

export type {
  AttachmentContent,
  AttachmentOptions,
  IAllureRuntime,
  ParameterOrString,
} from './runtime';
