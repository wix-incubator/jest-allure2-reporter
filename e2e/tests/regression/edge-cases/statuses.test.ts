import { allure } from '../../../../dist/api';

const dummyTest = () => expect(true).toBe(true);
const timeoutTest = (_done: Fn) => {
  expect({ seed: Math.random() }).toMatchSnapshot();
  // _done() is never called
};
const passingAssertion = () => expect(2 + 2).toBe(4);
const failingAssertion = () => expect(2 + 2).toBe(5);
const failingSnapshot = () => expect({ seed: Math.random() }).toMatchSnapshot();
const brokenAssertion = () => { throw new Error('This assertion is broken'); };

type Fn = () => any;

describe('Status tests', () => {
  describe('Simple', () => {
    test('passed', passingAssertion);
    test('failed assertion', failingAssertion);
    test('failed snapshot', failingSnapshot);
    test('failed timeout', timeoutTest, 100);
    test('broken', brokenAssertion);
    test.skip('skipped', passingAssertion);
    test.todo('todo');
  });

  describe.each([
    ['passing', passingAssertion],
    ['failed assertion in a ', failingAssertion],
    ['failed snapshot in a ', failingSnapshot],
    ['broken', brokenAssertion],
  ])('Status override in a %s', (_parentSuite, callback) => {
    describe.each([
      ['test', (callback: Fn) => (test('', callback), void 0)],
      ['test step', (callback: Fn) => (test('', () => allure.step('a step', callback)), void 0)],
      ['beforeEach hook', (callback: Fn) => (beforeEach(callback), test('', dummyTest), void 0)],
      ['afterEach hook', (callback: Fn) => (afterEach(callback), test('', dummyTest), void 0)],
      ...(callback === passingAssertion ? [
        ['beforeAll hook', (callback: Fn) => (beforeAll(callback), test('', dummyTest), void 0)],
        ['afterAll hook', (callback: Fn) => (afterAll(callback), test('', dummyTest), void 0)],
      ] as [string, (callback: Fn) => void][] : []),
    ])('%s', (_suite: string, hook: (callback: Fn) => void) => {
      hook(function () {
        allure.status('unknown', { message: 'Custom message', trace: 'Custom trace' });
        callback();
      });
    });
  });

  describe('Status override while timing out', () => {
    test('test', (done) => {
      timeoutTest(done);
      allure.status('unknown', {message: 'Custom message', trace: 'Custom trace'});
    }, 100);

    describe.each([
      ['beforeAll hook', beforeAll],
      ['beforeEach hook', beforeEach],
      ['afterEach hook', afterEach],
      ['afterAll hook', afterAll],
    ])('%s', (_suite, hook) => {
      hook(function (_done) {
        allure.status('unknown', {message: 'Custom message', trace: 'Custom trace'});
      }, 100);

      test('dummy', dummyTest);
    });
  });
});
