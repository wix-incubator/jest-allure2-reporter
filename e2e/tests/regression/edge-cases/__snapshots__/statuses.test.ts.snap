// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Status tests Simple failed snapshot 1`] = `
{
  "seed": 0.831296394914633,
}
`;

exports[`Status tests Simple failed timeout 1`] = `
{
  "seed": 0.3646202773486069,
}
`;

exports[`Status tests Status override in a failed snapshot in a  afterEach hook  1`] = `
{
  "seed": 0.5996337738043038,
}
`;

exports[`Status tests Status override in a failed snapshot in a  beforeEach hook  1`] = `
{
  "seed": 0.928365985442531,
}
`;

exports[`Status tests Status override in a failed snapshot in a  test  1`] = `
{
  "seed": 0.37722393964358303,
}
`;

exports[`Status tests Status override in a failed snapshot in a  test step  1`] = `
{
  "seed": 0.8442442168520534,
}
`;

exports[`Status tests Status override while timing out test 1`] = `
{
  "seed": 0.13785771624386745,
}
`;
